File: pledge-core/src/state.ts
// Pure-function state machine — SPEC §4.4.
//
// Transition table (paraphrased):
//
//   abandonment present      → broken (always; no honorable exit, §5.4)
//   contradictory outcomes   → disputed (§4.5)
//   outcome present          → outcome.outcome   (after dispute window passes)
//   now >= expires_at        → expired_unresolved
//   now >= resolves_at       → resolvable
//   else                     → pending
//
// `resolves_at` normalization (§4.4):
//   - time-typed: use the wall-clock timestamp directly.
//   - block-typed: use the wall-clock time of that block. If chain.blockTimes
//     names it, use that. Else if chain.tip_height >= block, conservatively
//     use chain.tip_time (the block has been mined). Else +infinity (the
//     block hasn't been mined yet → still pending).

import type {
    ClassifyStateInput,
    OutcomeEnvelope,
    PledgeEnvelope,
    PledgeState,
} from './types.js';

export function classifyState(input: ClassifyStateInput): PledgeState {
    const { pledge, outcome, abandonment, now, contradictoryOutcomes } = input;
    const nowMs = parseUtc(now);

    // Abandonment trumps everything: SPEC §5.4 — no honorable-exit path.
    if (abandonment !== null) {
        // Trust the caller to have signature-verified the abandonment via
        // verifyAbandonment(); state classification is a pure function of
        // valid inputs.
        return 'broken';
    }

    // Contradictory outcome envelopes from authorized resolvers → disputed.
    if (outcome !== null && contradictoryOutcomes && contradictoryOutcomes.length > 0) {
        for (const other of contradictoryOutcomes) {
            if (other.pledge_id === outcome.pledge_id && other.outcome !== outcome.outcome) {
                return 'disputed';
            }
        }
    }

    if (outcome !== null) {
        // Note: outcome.outcome can be 'expired_unresolved' or 'disputed' too;
        // the state machine just returns whatever the resolver classified.
        // The dispute_window_ends_at gate is informational here — verifiers
        // wanting "provisional vs final" distinction should branch on it
        // themselves; we return the eventual classification.
        return outcome.outcome;
    }

    const expiresAtMs = parseUtc(pledge.expires_at);
    if (nowMs >= expiresAtMs) {
        return 'expired_unresolved';
    }

    const resolvesAtMs = normalizeResolvesAt(pledge, input.chain);
    if (resolvesAtMs !== null && nowMs >= resolvesAtMs) {
        return 'resolvable';
    }

    return 'pending';
}

function normalizeResolvesAt(
    pledge: PledgeEnvelope,
    chain?: ClassifyStateInput['chain'],
): number | null {
    const r = pledge.resolves_at;
    if ('time' in r) return parseUtc(r.time);
    if ('block' in r) {
        if (!chain) return null; // +infinity — pledge stays pending without chain context
        if (chain.blockTimes && chain.blockTimes[r.block]) {
            return parseUtc(chain.blockTimes[r.block]!);
        }
        if (chain.tip_height >= r.block) {
            // Block has been mined; we don't know its exact time but we know
            // it's been mined by `tip_time`. Conservative: use tip_time as an
            // upper bound on the resolves moment. Verifiers wanting a precise
            // wall-clock SHOULD inject blockTimes[r.block].
            return parseUtc(chain.tip_time);
        }
        return null; // not yet mined
    }
    return null;
}

/**
 * Parse a strict ISO 8601 UTC timestamp (YYYY-MM-DDTHH:MM:SSZ) into epoch ms.
 * No fractional seconds, no offsets other than `Z` — SPEC §0.
 */
function parseUtc(s: string): number {
    const m = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})Z$/.exec(s);
    if (!m) throw new Error(`invalid ISO 8601 UTC string: ${s}`);
    const [, y, mo, d, h, mi, se] = m;
    return Date.UTC(
        Number(y),
        Number(mo) - 1,
        Number(d),
        Number(h),
        Number(mi),
        Number(se),
    );
}

/**
 * Detect whether two outcome envelopes for the same pledge are contradictory.
 * Two outcomes contradict when they reference the same pledge_id AND classify
 * to different outcome values (kept vs broken etc).
 */
export function outcomesContradict(a: OutcomeEnvelope, b: OutcomeEnvelope): boolean {
    if (a.pledge_id !== b.pledge_id) return false;
    return a.outcome !== b.outcome;
}

File: sdk-py/orangecheck/pledge/state.py
"""
Pure-function state machine — SPEC §4.4.

Mirrors the TS SDK's classifyState in src/state.ts. Same transition rules:

    abandonment present      → broken (always; no honorable exit, §5.4)
    contradictory outcomes   → disputed (§4.5)
    outcome present          → outcome.outcome
    now >= expires_at        → expired_unresolved
    now >= resolves_at       → resolvable
    else                     → pending
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any, Literal, Optional

PledgeState = Literal[
    "pending",
    "resolvable",
    "kept",
    "broken",
    "disputed",
    "expired_unresolved",
]

_ISO_UTC_STRICT = re.compile(r"^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})Z$")


@dataclass(frozen=True)
class ChainState:
    tip_height: int
    tip_time: str
    block_times: Optional[dict[int, str]] = None


def classify_state(
    *,
    pledge: dict[str, Any],
    outcome: Optional[dict[str, Any]],
    abandonment: Optional[dict[str, Any]],
    now: str,
    chain: Optional[ChainState] = None,
    contradictory_outcomes: Optional[list[dict[str, Any]]] = None,
) -> PledgeState:
    """Classify the state of a pledge given (pledge envelope, optional outcome
    envelope, optional abandonment envelope, current time, optional chain
    state for block-typed resolves_at, optional contradictory outcomes).

    All envelope arguments are JSON-shaped dicts (matching the wire form).
    """
    now_ms = _parse_utc(now)

    # Abandonment trumps everything — SPEC §5.4 / WHY §H8.
    if abandonment is not None:
        return "broken"

    # Contradictory outcome envelopes from authorized resolvers → disputed.
    if outcome is not None and contradictory_outcomes:
        for other in contradictory_outcomes:
            if (
                other.get("pledge_id") == outcome.get("pledge_id")
                and other.get("outcome") != outcome.get("outcome")
            ):
                return "disputed"

    if outcome is not None:
        return outcome.get("outcome")  # type: ignore[return-value]

    expires_at_ms = _parse_utc(pledge["expires_at"])
    if now_ms >= expires_at_ms:
        return "expired_unresolved"

    resolves_at_ms = _normalize_resolves_at(pledge, chain)
    if resolves_at_ms is not None and now_ms >= resolves_at_ms:
        return "resolvable"

    return "pending"


def _normalize_resolves_at(
    pledge: dict[str, Any], chain: Optional[ChainState]
) -> Optional[int]:
    r = pledge["resolves_at"]
    if "time" in r:
        return _parse_utc(r["time"])
    if "block" in r:
        if chain is None:
            return None  # +infinity — pledge stays pending without chain context
        block = int(r["block"])
        if chain.block_times and block in chain.block_times:
            return _parse_utc(chain.block_times[block])
        if chain.tip_height >= block:
            # Block has been mined; conservative bound is tip_time.
            return _parse_utc(chain.tip_time)
        return None  # not yet mined
    return None


def _parse_utc(s: str) -> int:
    """Parse strict ISO 8601 UTC (YYYY-MM-DDTHH:MM:SSZ) into epoch ms.

    Mirrors the TS SDK's parseUtc — no fractional seconds, no offsets other
    than ``Z``. Avoids datetime / strptime to keep behaviour deterministic
    across CPython versions and locales."""
    m = _ISO_UTC_STRICT.fullmatch(s)
    if not m:
        raise ValueError(f"invalid ISO 8601 UTC string: {s!r}")
    y, mo, d, h, mi, se = (int(g) for g in m.groups())
    # epoch (1970-01-01) → ms via day-of-year arithmetic.
    days = _date_to_days(y, mo, d) - _date_to_days(1970, 1, 1)
    return ((days * 24 + h) * 3600 + mi * 60 + se) * 1000


def _date_to_days(year: int, month: int, day: int) -> int:
    """Convert (Y, M, D) to a serial day count using the Gregorian calendar."""
    # Standard civil-from-days algorithm rearranged to days-from-civil.
    if month <= 2:
        year -= 1
        month += 12
    era = year // 400 if year >= 0 else (year - 399) // 400
    yoe = year - era * 400
    doy = (153 * (month - 3) + 2) // 5 + day - 1
    doe = yoe * 365 + yoe // 4 - yoe // 100 + doy
    return era * 146097 + doe - 719468


def outcomes_contradict(a: dict[str, Any], b: dict[str, Any]) -> bool:
    """True iff two outcome envelopes for the same pledge classify differently."""
    if a.get("pledge_id") != b.get("pledge_id"):
        return False
    return a.get("outcome") != b.get("outcome")
